aws/resource_aws_appmesh_virtual_router.go

```diff
@@ -0,0 +1,159 @@
+		VirtualRouterName: aws.String(d.Get("name").(string)),
+		Spec:              expandAppmeshVirtualRouterSpec(d.Get("spec").([]interface{})),
+	}
+
+	log.Printf("[DEBUG] Creating App Mesh virtual router: %#v", req)
+	resp, err := conn.CreateVirtualRouter(req)
+	if err != nil {
+		return fmt.Errorf("error creating App Mesh virtual router: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.VirtualRouter.Metadata.Uid))
+
+	return resourceAwsAppmeshVirtualRouterUpdate(d, meta)
+}
+
+func resourceAwsAppmeshVirtualRouterUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	if !d.IsNewResource() && d.HasChange("spec") {
+		_, v := d.GetChange("spec")
+		req := &appmesh.UpdateVirtualRouterInput{
+			MeshName:          aws.String(d.Get("mesh_name").(string)),
+			VirtualRouterName: aws.String(d.Get("name").(string)),
+			Spec:              expandAppmeshVirtualRouterSpec(v.([]interface{})),
+		}
+
+		log.Printf("[DEBUG] Updating App Mesh virtual router: %#v", req)
+		_, err := conn.UpdateVirtualRouter(req)
+		if err != nil {
+			return fmt.Errorf("error updating App Mesh virtual router: %s", err)
+		}
+	}
+
+	return resourceAwsAppmeshVirtualRouterRead(d, meta)
+}
+
+func resourceAwsAppmeshVirtualRouterRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	resp, err := conn.DescribeVirtualRouter(&appmesh.DescribeVirtualRouterInput{
+		MeshName:          aws.String(d.Get("mesh_name").(string)),
+		VirtualRouterName: aws.String(d.Get("name").(string)),
+	})
+	if err != nil {
+		if isAWSErr(err, "NotFoundException", "") {
+			log.Printf("[WARN] App Mesh virtual router (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading App Mesh virtual router: %s", err)
+	}
+	if aws.StringValue(resp.VirtualRouter.Status.Status) == appmesh.VirtualRouterStatusCodeDeleted {
+		log.Printf("[WARN] App Mesh virtual router (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", resp.VirtualRouter.VirtualRouterName)
+	d.Set("mesh_name", resp.VirtualRouter.MeshName)
+	d.Set("arn", resp.VirtualRouter.Metadata.Arn)
+	d.Set("created_date", resp.VirtualRouter.Metadata.CreatedAt.Format(time.RFC3339))
+	d.Set("last_updated_date", resp.VirtualRouter.Metadata.LastUpdatedAt.Format(time.RFC3339))
+	if err := d.Set("spec", flattenAppmeshVirtualRouterSpec(resp.VirtualRouter.Spec)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsAppmeshVirtualRouterDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	log.Printf("[DEBUG] Deleting App Mesh virtual router: %s", d.Id())
+	_, err := conn.DeleteVirtualRouter(&appmesh.DeleteVirtualRouterInput{
+		MeshName:          aws.String(d.Get("mesh_name").(string)),
+		VirtualRouterName: aws.String(d.Get("name").(string)),
+	})
+	if err != nil {
+		if isAWSErr(err, "NotFoundException", "") {
```

Comment:
Nit: SDK provided constant available:

```suggestion
		if isAWSErr(err, appmesh.ErrCodeNotFoundException, "") {
```